### ISSUE TITLE
Window is not responding after search&replace in 1000s files, survives reboot rendering VSCode unusable

### ISSUE THREAD

**Issue by bartekn-chromium** [microsoft/vscode#130634]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.58.0 2d23c42a936db1c7b3b06f918cde29561cc47cd6 x64
- OS Version: Debian GNU/Linux 11 (rodete)

Steps to Reproduce:
1. Do a bulk search&replace on a few thousand files, in a remote workspace

I put my VSCode in a permanently bad state, with the step above. Search&replace automatically saves the files, and it was taking long. At one point of time "The windows is not responding" popped up. When I clicked "Keep Waiting" the same file were attempted to be saved with on progress. So eventually I killed it and restarted, but "The windows is not responding" keeps popping up on startup and there is nothing I can do, not even close the workspace or open another one. Reset doesn't help either.

I tried --prof-startup, but "The windows is not responding" doesn't show up in this mode. When I kill it manually, profile has only 1 line and is useless.

I ran it with --vebose though and got this:
```
[175637:0812/112420.339455:INFO:CONSOLE(60)] "    at Ro.registerListeners (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:63957)
    at new m (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:63679)
    at new f (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1896:43352)
    at new Ro (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1896:66835)
    at g._createInstance (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:625:1371)
    at g.createInstance (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:625:790)
    at Object.createFileEditorInput (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2650:26118)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:30547
    at x.createOrGetCached (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:31401)
    at x.createEditorInput (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:30444)
    at Object.createEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:7690)
    at a.restoreBackups (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:5033)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112421.838779:INFO:CONSOLE(60)] "[b45] potential listener LEAK detected, having 1671 listeners already. MOST frequent listener (1493):", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112421.838830:INFO:CONSOLE(60)] "    at Ri.registerEditorListeners (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:73861)
    at Ri.openEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:73370)
    at Ii.openEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:15765)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:17624
    at Array.map (<anonymous>)
    at Ii.openEditors (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:17499)
    at async Promise.all (index 0)
    at async Object.U [as settled] (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:68:22106)
    at async x.openEditors (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:26806)
    at async a.restoreBackups (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:5072)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112422.484734:INFO:CONSOLE(60)] "[7dh] potential listener LEAK detected, having 1671 listeners already. MOST frequent listener (1494):", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112422.484774:INFO:CONSOLE(60)] "    at new l (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:171:31665)
    at new f (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:171:34118)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:171:41628
    at Array.forEach (<anonymous>)
    at g.push (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:171:41294)
    at Gt.redrawTab (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:21583)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:21280
    at Gt.doWithTab (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:13910)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:13664
    at Array.forEach (<anonymous>)
    at Gt.forEachTab (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:13643)
    at Gt.redraw (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:21247)
    at Gt.openEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2491:11414)
    at Ii.doShowEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:16252)
    at Ii.openEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:15790)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:17624
    at Array.map (<anonymous>)
    at Ii.openEditors (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2567:17499)
    at async Promise.all (index 0)
    at async Object.U [as settled] (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:68:22106)
    at async x.openEditors (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:26806)
    at async a.restoreBackups (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:5072)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[main 2021-08-12T02:24:26.980Z] menubar#runActionInRenderer { type: 'commandId', commandId: 'workbench.action.showAboutDialog' }
[175637:0812/112430.299597:INFO:CONSOLE(60)] "[6gg] potential listener LEAK detected, having 8268 listeners already. MOST frequent listener (7874):", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112430.299639:INFO:CONSOLE(60)] "    at Ro.registerListeners (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:63957)
    at new m (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:63679)
    at new f (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1896:43352)
    at new Ro (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1896:66835)
    at g._createInstance (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:625:1371)
    at g.createInstance (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:625:790)
    at Object.createFileEditorInput (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2650:26118)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:30547
    at x.createOrGetCached (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:31401)
    at x.createEditorInput (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:30444)
    at Object.createEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:7690)
    at a.restoreBackups (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:5033)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112439.739226:INFO:CONSOLE(60)] "[6gg] potential listener LEAK detected, having 8270 listeners already. MOST frequent listener (7874):", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
[175637:0812/112439.739264:INFO:CONSOLE(60)] "    at Ro.registerListeners (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:63957)
    at new m (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1502:63679)
    at new f (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1896:43352)
    at new Ro (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:1896:66835)
    at g._createInstance (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:625:1371)
    at g.createInstance (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:625:790)
    at Object.createFileEditorInput (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2650:26118)
    at vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:30547
    at x.createOrGetCached (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:31401)
    at x.createEditorInput (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2405:30444)
    at Object.createEditor (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:7690)
    at a.restoreBackups (vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js:2354:5033)", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (60)
```

**Comment by bartekn-chromium** [comment]
FYI I renamed ~/.config/Code and got unblocked. I kept the old directory so can help you troubleshoot.

**Comment by roblourens** [comment]
Doing a replace opens documents for all the replacements, but we should be trying to restore them since they weren't opened in editors right?

**Comment by bpasero** [comment]
@sandy081 I find it a bit surprising that we have no bound in the replace service for the edits that are applied. I would strongly suggest to split the work up into chunks of maybe 50-100 models each and then apply them one by one and only continue with the next chunk when the one chunk is finished.

Since we nicely show progress I do not think that is a big issue. In addition we could actually offer to cancel the operation.

**Comment by sandy081** [comment]
> I would strongly suggest to split the work up into chunks of maybe 50-100 models each and then apply them one by one and only continue with the next chunk when the one chunk is finished.

ReplaceService using BulkEditService to perform bulk edits. Is not the BulkEditService shall do it in chunks instead of its callers?

**Comment by bpasero** [comment]
Yeah maybe, cc @jrieken maybe this has come up before: basically whether the bulk edit service could split up the work of resolving models and applying edits in chunks to reduce pressure on memory and CPU.

Independent I am investigating the crash/hang when thousands of editors open. I notice how the tabs widget is not very clever when opening a lot of editors at once. I think I can fix that particular issue but it will probably not fix the issue of dirty editors opening.

**Comment by jrieken** [comment]
/duplicate of https://github.com/microsoft/vscode/issues/64485

The challenge is that we open all documents at once and that we check for unwanted concurrent modifications. If we replace in chunks than that's not easily possible anymore. Since there is no document to emit events and since changes might have been applied when encountering an invalid state

**Comment by bpasero** [comment]
https://github.com/microsoft/vscode/commit/77d32e443e0cbc7637c51e54e806a92ac7413d75 fixes the freeze/crash when thousands of editors open at once. My measurements improve opening 1600 files from 90s to 10s by optimizing the tabs control to only update after all tabs have been opened.
